pkg/storegateway/bucket_index_postings.go

```diff
@@ -19,64 +19,126 @@ import (
-// If addAll is set: special All postings minus postings for removeKeys labels. No need to merge postings for addKeys in this case.
-// If addAll is not set: Merge of postings for "addKeys" labels minus postings for removeKeys labels
-// This computation happens in ExpandedPostings.
-type postingGroup struct {
-	addAll     bool
-	addKeys    []labels.Label
-	removeKeys []labels.Label
+// rawPostingGroup keeps posting keys for single matcher. It is raw because there is no guarantee
+// that the keys in the gorup have a corresponding postings list in the index.
+// Logical result of the group is:
+// If isLazy == true: keys will be empty and lazyMatcher will be non-nil. Call prepare() to populate the keys.
+// If isSubtract == true: special All postings minus postings for keys labels.
+// If isSubtract == false: merge of postings for keys labels.
+// This computation happens in toPostingGroups.
+type rawPostingGroup struct {
+	isSubtract bool
+	labelName  string
+	keys       []labels.Label
+
+	isLazy      bool
+	lazyMatcher func(string) bool
+}
+
+func newRawPostingGroup(isSubtract bool, labelName string, keys []labels.Label) rawPostingGroup {
+	return rawPostingGroup{
+		isSubtract: isSubtract,
+		labelName:  labelName,
+		keys:       keys,
+	}
+}
+
+func newLazyPostingGroup(isSubtract bool, labelName string, matcher func(string) bool) rawPostingGroup {
+	return rawPostingGroup{
+		isLazy:      true,
+		isSubtract:  isSubtract,
+		labelName:   labelName,
+		lazyMatcher: matcher,
+	}
 }
 
-func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
-	return &postingGroup{
-		addAll:     addAll,
-		addKeys:    addKeys,
-		removeKeys: removeKeys,
+func (g rawPostingGroup) prepare(r indexheader.Reader) (postingGroup, error) {
+	var keys []labels.Label
+	if g.isLazy {
+		vals, err := r.LabelValues(g.labelName, g.lazyMatcher)
+		if err != nil {
+			return postingGroup{}, err
+		}
+		keys = make([]labels.Label, len(vals))
+		for i := range vals {
+			keys[i] = labels.Label{Name: g.labelName, Value: vals[i]}
+		}
+	} else {
+		var err error
+		keys, err = g.filterKeys(r)
+		if err != nil {
+			return postingGroup{}, errors.Wrap(err, "filter posting keys")
+		}
 	}
+
+	return postingGroup{
+		isSubtract: g.isSubtract,
+		keys:       keys,
+	}, nil
 }
 
-// NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
-func toPostingGroup(lvr indexheader.Reader, m *labels.Matcher) (*postingGroup, error) {
+func (g rawPostingGroup) filterKeys(r indexheader.Reader) ([]labels.Label, error) {
+	keys := g.keys
+	existingKeys := 0
+	for i, l := range keys {
+		if _, err := r.PostingsOffset(l.Name, l.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
+			// This label name and value doesn't exist in this block, so there are 0 postings we can match.
+			// Set it to an empty value, so we can filter it out later.
+			keys[i] = labels.Label{}
```

Comment:
Is it safe to override `g.keys`? `keys := g.keys` doesn't copy the underlying slice.

Reply:
If it's safe to modify `g.keys` then we could also modify it in-place instead of creating `filtered` below.

Reply:
yes, i can do that